engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/aggregate/aggregates.scala

```diff
@@ -171,6 +172,75 @@ object aggregates {
 
   }
 
+  object CountWhenAggregator extends Aggregator {
+
+    override type Element = java.lang.Boolean
+
+    override type Aggregate = java.lang.Long
+
+    override def zero: java.lang.Long = 0L
+
+    override def addElement(element: java.lang.Boolean, aggregate: Aggregate): Aggregate =
+      if (element) aggregate + 1 else aggregate
+
+    override def mergeAggregates(aggregate1: Aggregate, aggregate2: Aggregate): Aggregate = aggregate1 + aggregate2
+
+    override def result(finalAggregate: Aggregate): AnyRef = finalAggregate
+
+    override def computeOutputType(input: typing.TypingResult): Validated[String, typing.TypingResult] = {
+      if (input.canBeSubclassOf(Typed[Boolean])) {
+        Valid(Typed[Long])
+      } else {
+        Invalid(s"Invalid aggregate type: ${input.display}, should be: ${Typed[Boolean].display}")
+      }
+    }
+
+    override def computeStoredType(input: typing.TypingResult): Validated[String, typing.TypingResult] =
+      computeOutputType(input)
+  }
+
+  object AverageAggregator extends Aggregator {
+
+    override type Element = java.lang.Number
+
+    override type Aggregate = AverageAggregatorState
+
+    override def zero: AverageAggregatorState = AverageAggregatorState(null, 0)
+
+    override def addElement(element: java.lang.Number, aggregate: Aggregate): Aggregate =
+      AverageAggregatorState(MathUtils.largeFloatingSum(element, aggregate.sum), aggregate.count + 1)
+
+    override def mergeAggregates(aggregate1: Aggregate, aggregate2: Aggregate): Aggregate =
+      AverageAggregatorState(
+        MathUtils.largeFloatingSum(aggregate1.sum, aggregate2.sum),
+        aggregate1.count + aggregate2.count
+      )
+
+    override def result(finalAggregate: Aggregate): AnyRef = {
+      val count = finalAggregate.count
+      finalAggregate.sum match {
+        // TODO_PAWEL is it ok? there is no nan for BigDecimal, only null. But on the other hand I would expect Double.nan if result type is Double
+        case null                      => null
+        case sum: java.lang.Double     => Predef.double2Double(sum / count)
+        case sum: java.math.BigDecimal => (BigDecimal(sum) / BigDecimal(count)).bigDecimal
+      }
+    }
+
+    override def computeOutputType(input: typing.TypingResult): Validated[String, typing.TypingResult] = {
+
+      if (!input.canBeSubclassOf(Typed[Number])) {
+        Invalid(s"Invalid aggregate type: ${input.display}, should be: ${Typed[Number].display}")
+      } else {
+        Valid(ForLargeFloatingNumbersOperation.promoteSingle(input))
+      }
+    }
+
+    override def computeStoredType(input: typing.TypingResult): Validated[String, typing.TypingResult] =
```

Comment:
Uhh, I forgot that we have this clever mechanism, that generates flink's type information based on the `TypingResult`. Take a look at `TypingResultAwareTypeInformationDetection`. As far I remember, it was important:
1. For the time optimization reasons
2. To make state readable across deployments - I mean, if the state is stored using wrong TypeInformation, features like redeploy could not work. To redeploy you need to hit the deploy button for already running scenario.

I think that we should consider:
- Doing some tests around this
- Changing the type into Map[String, Any] and typing result to `TypedObjectTypingResult` with two fields: sum and count. The type of `sum` should be the same as returned by `computeOutputType`

Reply:
Łukasz Jędrzejewski suggested that in esp they add @TypeInfo  to case classes to speed up the serialization. I did it in commit https://github.com/TouK/nussknacker/pull/5313/commits/7abb8bdce557805262831d3603d99feeefca9b6f . Does this solve this problem?

Reply:
It almost handles it correctly. I checked in the debugger such a code:
```
TypeInformation.of(classOf[AverageAggregatorState])
```
and I see that one problem that remain is that `fieldType` for sum is still generic.
We can easily fix that by replacing the `sum` with nullable `sumDouble` and `sumBigDecimal`. Unfortunatelly, `CaseClassTypeInfoFactory` doesn't handles either sum types (`Either`, sealed traits) or `Option`.